### ISSUE TITLE
[Scope Change Request] page 502 OAuth2ControlAddIn 

### ISSUE THREAD

**Issue by tinestaric** [microsoft/ALAppExtensions#21753]
Hey,

is there any specific reason why procedures in the OAuth2ControlAddIn page are marked with OnPrem Scope? 

I'm trying to reuse the page when integrating with a 3rd party identity provider, but due to the scopes, I need to copy the whole object.

additionally, the GetOAuthProperties procedure of the codeunit 502 OAuth2Impl has too simple parsing of fragments in the query string:

``` AL
    [NonDebuggable]
    procedure GetOAuthProperties(AuthorizationCode: Text; var CodeOut: Text; var StateOut: Text; var AdminConsent: Text)
    begin
        if AuthorizationCode = '' then begin
            Session.LogMessage('0000C1V', AuthorizationCodeErr, Verbosity::Error, DataClassification::SystemMetadata, TelemetryScope::ExtensionPublisher, 'Category', Oauth2CategoryLbl);
            exit;
        end;

        if AuthorizationCode.EndsWith('#') then
            AuthorizationCode := CopyStr(AuthorizationCode, 1, StrLen(AuthorizationCode) - 1);

        CodeOut := GetPropertyFromCode(AuthorizationCode, 'code');
        StateOut := GetPropertyFromCode(AuthorizationCode, 'state');
        AdminConsent := GetPropertyFromCode(AuthorizationCode, 'admin_consent');
    end;
```

I know AAD by default doesn't include any fragments, but a lot of IdP's do. I'm not familiar with how you operate with internal implementation codeunits, do you provide events within them? If yes, then we could resolve the fragments on our own. An alternative would be an event before the call to GetOAuthProperties from the OAuth2ControlAddIn page.

If the answer to any of the two requests is that you only intend to support OAuth flows with AAD, then we'll just continue using the copied object. 
Internal work item: [AB#538510](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/538510)

**Comment by pri-kise** [comment]
@tinestaric you cannot subscribe to events of internal objects.
Events would have to be added in a public object and being called from the internal object.

Maybe you can elaborate your needed changes in more details, that someone of microsoft can decide if you can create a pull request to improve the current situation.

**Comment by tinestaric** [comment]
What I'm trying to do is implement OAuth2 Auth Code flow, but using a different Identity Provider than AAD. Specifically, I was working with supporting Facebook. 

The Public procedures of the OAuth2 codeunit are tightly coupled with AAD's flows, so I started implementing my own but wanted to reuse the same Control Add-In that OAuth2 module uses, as the part of the flow where a pop-up is raised and the query string is retrieved from the Redirect Url is basically the same for any OAuth2 implementation.

The two blockers I've hit are that procedure on OAuth2ControlAddIn page are all marked with OnPrem Scope, and the way Query string is parsed once retrieved from  the Redirect URL

```AL
    [Scope('OnPrem')]
    [NonDebuggable]
    procedure SetOAuth2Properties(AuthRequestUrl: Text; AuthInitialState: Text)
    begin
        OAuthRequestUrl := AuthRequestUrl;
        State := AuthInitialState;
    end;

    [Scope('OnPrem')]
    [NonDebuggable]
    procedure GetAuthCode(): Text
    begin
        exit(AuthCode);
    end;

    [Scope('OnPrem')]
    [NonDebuggable]
    procedure GetAuthError(): Text
    begin
        exit(AuthError);
    end;

    [Scope('OnPrem')]
    [NonDebuggable]
    procedure GetGrantConsentSuccess(): Boolean
    begin
        exit(HasAdminConsentSucceded);
    end;
```

I couldn't really see why the procedures are scoped like this, as I can just copy the whole object, and it will work just fine in a Cloud scope. So my first question is, is there a specific reason to have the OnPrem scope that I'm not seeing?


The second blocker was the fragment parsing that is done in the GetOAuthProperties. Specifically, in the case of Facebook, the query returned to the redirect URL is something like https://client.example.com/callback?code=g23flv...&state=32cvlse...#_=_

The code of the GetOAuthProperties then incorrectly parsed the state query parameter as "32cvlse...#_=_", I didn't exclude the fragment part (as the code only looks if the "#" char is at the end of the string).

The approach I took with my implementation was to build a URI and use its GetQuery() procedure to correctly get only the Query string.

```AL
        Uri.Init('https://client.example.com/callback?' + AuthorizationCode);
        AuthorizationCode := Uri.GetQuery().TrimStart('?');
```

As extensibility option, there could simply be an event before the GetOAuthProperties on the OAuth2ControlAddIn and I could parse the Query string depending on my specific needs.

```
                trigger AuthorizationCodeRetrieved(code: Text)
                var
                    StateOut: Text;
                    AdminConsentTxt: Text;
                begin
                    OnBeforeGetOAuthProperties(Handled, code, PropertiesOut);
                    if not Handled then
                            OAuth2Impl.GetOAuthProperties(code, AuthCode, StateOut, AdminConsentTxt);
                    ...
```

But mainly what I'm looking for is an answer if I should treat the OAuth2ControlAddIn page as Internal, where Microsoft only supports the AAD OAuth2 flows, and we should implement (copy) the Page on our own if we want to support 3rd party identity providers.
